system/threads.cpp

```diff
@@ -115,44 +166,155 @@ long __syscall_futex(uint32_t* uaddr, int futex_op, ...)
+							int64_t timeNow = timeNowStruct.tv_sec * 1000000000 + timeNowStruct.tv_nsec;
+							if (timeNow - startTime >= timeout)
+							{
+								removeNodeFromFutexList(&futexNode);
+								return -ETIMEDOUT;
+							}
+						}
+					}
+					futexNode.sleepAddress = futexNode.nextSleepAddress;
+					futexNode.nextSleepAddress = nullptr;
+					futexSpinLock.lock();
+					mainThreadWaitAddress.store(futexNode.sleepAddress);
+					futexSpinLock.unlock();
+				}
+				removeNodeFromFutexList(&futexNode);
+			}
+			else
+			{
+				addNodeToFutexList(&futexNode);
+				uint32_t ret = __builtin_cheerp_atomic_wait(uaddr, val, timeout);
+				if (ret == 1)
+				{
+					removeNodeFromFutexList(&futexNode);
+					return -EAGAIN; // Value at uaddr did not match val.
+				}
+				else if (ret == 2)
 				{
-					struct timespec startTimeStruct;
-					clock_gettime(CLOCK_MONOTONIC, &startTimeStruct);
-					startTime = startTimeStruct.tv_sec * 1000000000 + startTimeStruct.tv_nsec;
+					removeNodeFromFutexList(&futexNode);
+					return -ETIMEDOUT;
 				}
-				while (mainThreadWaitAddress.load() != 0)
+				while (futexNode.nextSleepAddress != nullptr)
 				{
+					futexNode.sleepAddress = futexNode.nextSleepAddress;
+					futexNode.nextSleepAddress = nullptr;
+					// We need to recalculate the timeout.
 					if (timeout != -1)
 					{
 						struct timespec timeNowStruct;
 						clock_gettime(CLOCK_MONOTONIC, &timeNowStruct);
 						int64_t timeNow = timeNowStruct.tv_sec * 1000000000 + timeNowStruct.tv_nsec;
-						if (timeNow - startTime >= timeout)
-							return ETIMEDOUT;
+						int64_t timePassed = timeNow - startTime;
+						if (timePassed >= timeout)
+						{
+							removeNodeFromFutexList(&futexNode);
+							return -ETIMEDOUT;
+						}
+						else
+							timeout -= timePassed;
+					}
+					// This new wait is caused by a requeue. We do not know the value at this new address.
+					// A return value of 1 therefore now means we need to try again with the new value until we succeed.
+					uint32_t ret = 1;
+					while (ret == 1)
+						ret = __builtin_cheerp_atomic_wait(futexNode.sleepAddress, val, timeout);
+					if (ret == 2)
+					{
+						removeNodeFromFutexList(&futexNode);
+						return -ETIMEDOUT;
 					}
 				}
+				removeNodeFromFutexList(&futexNode);
 			}
-			else
+			return 0;
+		}
+		case FUTEX_REQUEUE:
+		case FUTEX_CMP_REQUEUE:
+		{
+			va_list ap;
+			va_start(ap, futex_op);
+			uint32_t val = va_arg(ap, uint32_t);
+			uint32_t val2 = va_arg(ap, unsigned long);
+			uint32_t* uaddr2 = va_arg(ap, uint32_t*);
+
+			if (futex_op == FUTEX_CMP_REQUEUE)
 			{
-				uint32_t ret = __builtin_cheerp_atomic_wait(uaddr, val, timeout);
-				if (ret == 1)
-					return EAGAIN; // Value at uaddr did not match val.
-				else if (ret == 2)
-					return ETIMEDOUT;
+				// This needs to happen atomically? (Again at the end?) // TODO
+				//
+				//
+				//
+				uint32_t val3 = va_arg(ap, uint32_t);
+				if (*uaddr != val3)
+				{
+					va_end(ap);
```

Comment:
can't we do va_end before the if so we do it only once?

Reply:
Keeping this as is.